doc/widget_server.md

```diff
@@ -0,0 +1,417 @@
+    0
+    <button onClick={() => handle({h:1, r:[51]}, 'onClick')}>-<button>
+</>
+```
+
+## Handling a widget event
+
+When the user interacts with the UI, it may trigger events (eg `onClick` in the above example).
+These events must be passed to the Lean server so that the UI can be updated. The procedure for this is as follows:
+
+1. Send an `WidgetEventRequest`:
+    ```ts
+    interface WidgetEventRequest extends Request {
+        command: 'widget_event';
+        /** The event kind that caused this request. */
+        kind: 'onClick' | 'onMouseEnter' | 'onMouseLeave' | 'onChange';
+        /** The handler provided by the `WidgetElement`'s event object. */
+        handler: WidgetEventHandler;
+        /** use `type:'string'` in the case of `onChange` otherwise use `'unit'` */
+        args: { type: 'unit' } | { type: 'string'; value: string };
+        file_name: string;
+        /** These should be the same as the WidgetIdentifier retrieved from the `get_widget` response. */
+        line: number; column: number; id: number;
+    }
+    ```
+2. Receive a `WidgetEventResponse`:
+   ```ts
+    type WidgetEffect =
+        | {kind: 'insert_text'; text: string}
+        | {kind: 'reveal_position'; file_name: string; line: number; column: number}
+        | {kind: 'highlight_position'; file_name: string; line: number; column: number}
+        | {kind: 'clear_highlighting'}
+        | {kind: 'copy_text'; text:string}
+        | {kind: 'custom'; key: string; value: string}
+
+    interface WidgetEventRecordSuccess {
+        status: 'success';
+        widget: {
+            line: number;
+            column: number;
+            id: number;
+            html : WidgetComponent
+        };
+        effects?: WidgetEffect[];
+    }
+
+    interface WidgetEventRecordInvalid {
+        status: 'invalid_handler';
+    }
+
+    interface WidgetEventRecordError {
+        status: 'error';
+        message: string;
+    }
+
+    type WidgetEventRecord =
+        | WidgetEventRecordSuccess
+        | WidgetEventRecordInvalid
+        | WidgetEventRecordError;
+
+    interface WidgetEventResponse extends CommandResponse {
+        record: WidgetEventRecord;
+    }
+
+   ```
+   Some branching cases:
+   - If the `WidgetEventRecord` has `status: 'error'`, the client should enter an error state and report the error to the user.
+   - If the `WidgetEventRecord` has `status: 'invalid'`, then the client should throw away the widget and try getting it again as detailed in the above section.
+   - Otherwise, continue to the next step.
+3. render the resulting `widget` object in `WidgetEventRecordSuccess` in place of the existing render.
+4. If `effects` is not undefined and non-empty, perform these effects in order (details in next section).
+
+### Effects
+
+When a `WidgetEventResponse` is returned, it may contain 'effects' these are changes to the editor state that the UI is requesting.
```

Comment:
```suggestion
When a `WidgetEventResponse` is returned, it may contain 'effects', which are changes to the editor state that the UI is requesting.
```